crates/circuit/src/dag_node.rs

```diff
@@ -0,0 +1,283 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::circuit_instruction::CircuitInstruction;
+use pyo3::prelude::*;
+use pyo3::types::{PyDict, PyList, PySequence, PyString, PyTuple};
+use pyo3::{intern, PyObject, PyResult};
+
+/// Parent class for DAGOpNode, DAGInNode, and DAGOutNode.
+#[pyclass(module = "qiskit._accelerate.circuit", subclass)]
+#[derive(Clone, Debug)]
+pub struct DAGNode {
+    #[pyo3(get, set)]
+    pub _node_id: isize,
+}
+
+#[pymethods]
+impl DAGNode {
+    #[new]
+    #[pyo3(signature=(nid=-1))]
+    fn new(nid: isize) -> Self {
+        DAGNode { _node_id: nid }
+    }
+
+    fn __getstate__(&self) -> isize {
+        self._node_id
+    }
+
+    fn __setstate__(&mut self, nid: isize) {
+        self._node_id = nid;
+    }
```

Comment:
Strictly these two could also be replaced with a `__getnewargs__` as well, but it's unimportant - I wouldn't be surprised if they're never called at all, since I don't think we ever construct bare `DAGNode`.